File: backend/app/api/schemas/preferences.py
"""Schemas for user preferences endpoints."""
from __future__ import annotations

from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class PersonalSlots(BaseModel):
    fitness: Literal["morning", "afternoon", "evening"] = "morning"
    learning: Literal["morning", "afternoon", "evening"] = "evening"
    admin: Literal["weekend", "evenings"] = "weekend"


class AvailabilityProfile(BaseModel):
    work_days: List[Literal["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]] = Field(
        default_factory=lambda: ["Mon", "Tue", "Wed", "Thu", "Fri"]
    )
    work_start: str = "09:00"
    work_end: str = "18:00"
    peak_energy: Literal["morning", "evening"] = "morning"
    work_mode_enabled: bool = False
    personal_slots: PersonalSlots = PersonalSlots()


class PreferencesResponse(BaseModel):
    user_id: UUID
    coaching_paused: bool
    weekly_plans_enabled: bool
    interventions_enabled: bool
    availability_profile: AvailabilityProfile
    request_id: str


class PreferencesUpdateRequest(BaseModel):
    user_id: UUID
    coaching_paused: Optional[bool] = None
    weekly_plans_enabled: Optional[bool] = None
    interventions_enabled: Optional[bool] = None
    availability_profile: Optional[AvailabilityProfile] = None

File: mobile/src/api/preferences.ts
import { apiRequest } from "./client";

export type AvailabilityProfile = {
  work_days: string[];
  work_start: string;
  work_end: string;
  peak_energy: "morning" | "evening";
  work_mode_enabled: boolean;
  personal_slots: {
    fitness: "morning" | "afternoon" | "evening";
    learning: "morning" | "afternoon" | "evening";
    admin: "weekend" | "evenings";
  };
};

export type PreferencesResponse = {
  user_id: string;
  coaching_paused: boolean;
  weekly_plans_enabled: boolean;
  interventions_enabled: boolean;
  availability_profile: AvailabilityProfile;
  request_id: string;
};

export async function getPreferences(userId: string): Promise<{ data: PreferencesResponse; requestId: string | null }>
{
  const { data, response } = await apiRequest<PreferencesResponse>(`/preferences?user_id=${userId}`);
  return { data, requestId: data.request_id || response.headers.get("X-Request-Id") };
}

export async function updatePreferences(
  userId: string,
  patch: Partial<
    Pick<PreferencesResponse, "coaching_paused" | "weekly_plans_enabled" | "interventions_enabled" | "availability_profile">
  >,
): Promise<{ data: PreferencesResponse; requestId: string | null }>
{
  const { data, response } = await apiRequest<PreferencesResponse>("/preferences", {
    method: "PATCH",
    body: { user_id: userId, ...patch },
  });
  return { data, requestId: data.request_id || response.headers.get("X-Request-Id") };
}
